### DISCUSSION TITLE
How to set the fps of cameracapture

### DISCUSSION THREAD

**Discussion by tudou2333333** [bonsai-rx/bonsai#1647] [Q&A]
I have a usb camera, and it can support 60hz with windows camera. But no matter how do I change the parameters of cameracapture in bonsai cameraproperties, I can only get 30hz video. 
![image](https://github.com/bonsai-rx/bonsai/assets/102028902/dff613cf-c721-49aa-b360-b4ecf66a0177)
I changed the parameters here. 
![捕获](https://github.com/bonsai-rx/bonsai/assets/102028902/e3aa99a1-02e9-4d89-8a52-80771e20eb0f)
And this is the timestamps of my recording. 
Any help would be appreciated!

**Comment by bruno-f-cruz** [comment]
`CameraCapture` uses an OpenCV backend to try to set this property. Several reasons this could fail, most common than not is the camera you are trying to control not honoring the full opencv standard.

Can you try to use VideoCaptureDevice instead?

**Reply by tudou2333333** [reply]
Thank you so much! The VideoCaptureDevice fixed my problem!

**Reply by tomasmanu** [reply]
Hi there! We're having the same problem with a PS3 Eye camera. We're able to set the stream at 60 fps with OBS Studio for instance, but when we want to interface the camera using Bonsai, we always have the limit of 30 fps, using `CameraCapture` and  also with `VideoCaptureDevice`. In fact, when we try to open the Default editor of `VideoCaptureDevice`, we can't as Bonsai would never open that window. How can we solve that @bruno-f-cruz ?

**Reply by bruno-f-cruz** [reply]
A few things to unpack here.
What drivers are you using to interface with PS3 Eye cameras?
What backend is OBS studio using to interface with the camera?
Are you sure the camera is acquiring at 60fps in OBS? is it possible that OBS is interpolating / duplicating frames?
VideoCaptureDevice does not have a default editor, not sure what you mean by this? (nvm it does have one, my camera just doesn't implement it... had to look into the source code)

**Reply by tomasmanu** [reply]
We use the driver from this repo https://github.com/jkevin/PS3EyeDirectShow/releases. It's the **PS3EyeInstallerBeta2.msi**
With respect to OBS, I don't quite know what you're referring to with what backend it's using. What is a backend precisely?
In what respects the acquisition, we checked different videos recorded with OBS at different fps by inspecting them frame by frame with FIJI, in order to see if there were duplicated frames. At 30 and 60 fps, there were no duplications, but if we set a higher frame rate (i.e 90 or 120), we could evidence duplications: at 90 fps, 1 of every 3 frames was a duplication, and at 120 fps, 1 of every 2 frames was a duplication.
I refer to the following as the default editor of `VideoCaptureDevice`: 
![image](https://github.com/bonsai-rx/bonsai/assets/69548554/5c0b1537-ef03-42f7-9b3f-6a5cc3d9cc69)
I thought the fps could be modified there, but I'm not quite sure now.

**Reply by bruno-f-cruz** [reply]
If i remember correctly, this UI is just the UI that the direct show driver of the camera implements. Different cameras will actually show different UIs here. I suspect the DirectShow driver does not implement a UI for the camera you have.
What options do you see under `Format` property? What is the resolution of the output image you have in bonsai?

**Reply by tomasmanu** [reply]
There are just two options under `Format` property: a higher and a lower resolution, both indicating this "(30-60 fps)". But none of them produce a 60 fps stream.
Do you know of another driver for PS3 Eye camera that works properly for this aim of better frame rate?

**Reply by bruno-f-cruz** [reply]
What are the resolution values? What happens if you set the lowest resolution and set the exposure to a very small value?
I unfortunately never used these cameras so no idea, sorry.

**Reply by tomasmanu** [reply]
But how do I change the exposure if the Default Editor of `VideoCaptureDevice` doesn't work for the driver I'm using for this camera? Is there another way of changing the exposure with this node or another one? 
I'm not currently at the lab so I can't check the precise resolutions that are available. Tomorrow I can check and reply again with that data.

**Reply by bruno-f-cruz** [reply]
To set exposure: `CaptureProperties` -> click `...` -> click `Add` -> click `ControlFlags` and set `Manual`, click `Property` and set it to `Exposure` set a value under `Value`. The units are sometimes arbitrary, you might need to play around with it and might even have to go to negative values if the units are logarithmic (weird I know.. https://www.kurokesu.com/main/2020/05/22/uvc-camera-exposure-timing-in-opencv/)
